DHFPROD-4308: Middle-Tier: Create extension to existing mapping artif…

hao1st opened the pull request:
https://project.marklogic.com/jira/browse/DHFPROD-4308

ryanjdew commented:
Looks good. In DH today, we're are using "targetEntity" instead of "entityName". 

@rjrudin I know you're big on good naming conventions. Thoughts? One thing to consider is that in the current mappings, while it will work with only specifying an entity model name, I specifically made it so you could specify the absolute IRI of an entity, so versioned entities could be used.

rjrudin commented on marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/artifacts/core.sjs:
```diff
@@ -56,11 +58,42 @@ function getArtifacts(artifactType) {
+function getArtifactsGroupbyEntity(artifactType, queriesForAnd) {
+    const artifacts = [];
+    const entityNames = getEntityNames();
+    for (const ename of entityNames) {
```
This is doing N queries, where N = entityNames.length. It's faster/better to do a single query - just pass entityNames into the cts.jsonPropertyValueQuery. Then iterate over the results with a map function. Can get it all done this simply (I often refer to https://www.w3schools.com/jsref/jsref_obj_array.asp for things like this):

```
function getArtifactsGroupByEntity(artifactType, queriesForAnd) {
  cts.search(cts.andQuery(queriesForAnd.concat(cts.jsonPropertyValueQuery("entityName", getEntityNames())).map(config => {
    const configByEntity = {
      name: config.path.to.entityName (I don't know the path)
    };
    switch(artifactType) {
    ....
    }
    return configByEntity;
  });
}
```

Note the use of "concat" to avoid modifying the queriesForAnd array that's passed in.

hao1st replied:
`let entities = cts.search(cts.collectionQuery("http://marklogic.com/entity-services/models"))
const entityNames = entities.toArray().map(e => e.xpath("//info//title"));
const res = cts.search(cts.andQuery(queriesForAnd.concat(cts.jsonPropertyValueQuery("entityName", entityNames)))).toArray();

let res2 = res
    .reduce((acc, mapping) => {
      acc[mapping.entityName] = acc[mapping.entityName] || {};
      acc[mapping.entityName].mapping = (acc[mapping.entityName].mappings || []);
      acc[mapping.entityName].mapping.push(mapping);
      return acc;
    }, {});`

I tried to use the above code, and did not work.

hao1st replied:
Fixed.

rjrudin commented on marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/artifacts/core.sjs:
```diff
@@ -56,11 +58,42 @@ function getArtifacts(artifactType) {
+        let configs = cts.search(cts.andQuery(queriesForAnd));
+        // artifactType can be mapping, mastering, merging
+        switch (artifactType) {
+            case "mapping":
```
Ideally, this service wouldn't have much knowledge of the concrete artifact types. Is it necessary to store the "configs" under different key names? Couldn't it just be "config" or "settings", regardless of the artifact type?

hao1st replied:
agree. we can use config. @ryanjdew do you agree?

rjrudin commented on marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/artifacts/core.sjs:
```diff
@@ -225,6 +258,16 @@ function deleteArtifactSettings(artifactType, artifactName, artifactVersion = 'l
 }
 
+function getEntityNames() {
+    let entities = cts.search(cts.collectionQuery("http://marklogic.com/entity-services/models"))
```
You can use "map" here to cut out a good chunk of code:

    cts.search(...).toArray().map(e => e.xpath("//info//title");

rjrudin commented on one-ui/src/main/java/com/marklogic/hub/curation/controllers/CollectionsController.java:
```diff
@@ -0,0 +1,4 @@
+package com.marklogic.hub.curation.controllers;
+
+public class CollectionsController {
```
This seems like it can be deleted?

hao1st replied:
my bad, something wrong during commit and push.

rjrudin replied:
This is still in the PR, it should be removed, right?

rjrudin replied:
Let's put this in a separate PR for a separate sub-task under 3991. It needs a test for it, and we should instead use a new DS endpoint instead of continuing use of REST endpoints (and the mlCollections REST extension is doing too many things too, it's not just used for retrieving collections).

rjrudin commented on one-ui/src/main/java/com/marklogic/hub/curation/services/CollectionsService.java:
```diff
@@ -0,0 +1,76 @@
+import java.util.List;
+
+@Service
+public class CollectionsService {
```
This doesn't appear to be used anywhere, so it should be deleted.

If there is a need for it, I'd like to talk about it. There is nothing preventing a database from having millions of collections or more. And collections are implementation details - it's very rare that we ever want to expose them to users. So I'd want to think carefully about the requirement of "Get me all the collections except 'internal' ones".

hao1st commented:
> Looks good. In DH today, we're are using "targetEntity" instead of "entityName".
> 
> @rjrudin I know you're big on good naming conventions. Thoughts? One thing to consider is that in the current mappings, while it will work with only specifying an entity model name, I specifically made it so you could specify the absolute IRI of an entity, so versioned entities could be used.
> 
> And apologies @hao1st, I should have caught that when reviewing the ticket before.

@ryanjdew NP, I changed it and modified the ticket

bsrikan commented on marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/artifacts/mapping.sjs:
```diff
@@ -0,0 +1,78 @@
+    return artifact;
+}
+
+module.exports = {
```
@hao1st can we please add a test for validateArtifact?

hao1st replied:
i will talk to @ryanjdew tomorrow, and we can simply the validateArtifact implemented in core.sjs. Actually we do not need to have this one in the artifact module.

hao1st replied:
@ryanjdew In the core.sjs, 
```
function validateArtifact(artifactType, artifactName, artifact) {
    const artifactLibrary =  getArtifactTypeLibrary(artifactType);
    const validatedArtifact = artifactLibrary.validateArtifact(artifact, artifactName);
    if (validatedArtifact instanceof Error) {
        returnErrToClient(400, 'BAD REQUEST', validatedArtifact.message);
    }
    return validatedArtifact;
}

I think we can move the function used in loadData.sjs to core.sjs

```
function validateArtifact(artifact) {
    const missingProperties = requiredProperties.filter((propName) => !artifact[propName]);
    if (missingProperties.length) {
        return new Error(`Missing the following required properties: ${JSON.stringify(missingProperties)}`);
    }
    return artifact;
}
```
Different artifacts just need to expose "requiredProperties" so that we do not need to override/duplicate validateArtifact with the same implmentation.

ryanjdew replied:
@hao1st  We'll want to extend the validation to be able to encompass more than the required properties. 

e.g., for mapping, we'll want to enable the validation to eventually do the validation XPath expressions, etc. We already have the libraries for that from what we used in QS. We'll want to eventually call those from the mapping artifact library.

hao1st replied:
@bsrikan OK, I added the validation test case for mapping test.

bsrikan commented on one-ui/src/main/java/com/marklogic/hub/curation/services/CollectionsManager.java:
```diff
@@ -0,0 +1,75 @@
+import org.springframework.stereotype.Service;
+
+@Service
+public class CollectionsManager {
```
Does this go with the discussion we had today. Should this be @Service or @Component.

rjrudin replied:
I don't think this class is supposed to be in here, right? It's only used by CollectionsCOntroller, which isn't used - nor seems relevant to DHFPROD-4308.

rjrudin commented on marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/artifacts/core.sjs:
```diff
@@ -26,7 +27,8 @@ const cachedArtifacts = {};
     flows: Flows,
-    stepDefinitions: StepDefs
+    stepDefinitions: StepDefs,
+    mapping: Mapping
```
For consistency, shouldn't this be "mappings" and "Mappings"?

hao1st replied:
@rjrudin @ryanjdew We used to have a folder like "mappings/{mappingstepname}" to be used to store [entity model <=> doc instance] mapping configuration with a filename like {mappingstepname}.{version}.mapping.json .
currently if i use "mapping" as artifact type, a new "mapping" folder willl be created to store the mapping artifacts with filename  like {mappingName}.mapping.json. If we use "mapings", it will cause name conflict. I am not sure it is a good name or a better folder stucture we should define for "mapping" operations.

rjrudin replied:
I'm confused - the directory for mappings in DHF today is "mappings" (and "flows" for flows and "step-definitions" for step definitions). Aren't these mapping artifacts still going into the "mappings" directory?

hao1st replied:
Yes, It can.  we have {mappingName} as subfolder. i will modify it to mappings

bsrikan replied:
The dir structure today for mappings is mappings/<flowName-stepName>/<flowName-stepName-version>.mapping.json

version starts with 0 when a mapping step is created and bumped to 1 when mapping is configured after ingestion. Version remains at 1 with further changes to mapping.

rjrudin commented on marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/artifacts/core.sjs:
```diff
@@ -49,19 +51,31 @@ function getTypesInfo() {
+    return [];
+}
+
+function getArtifactsGroupByEntity(queriesForAnd) {
```
Is this returning artifacts, or configs, or settings? Are "configs" the same as "settings"? I don't think this is returning artifacts, which are the actual flows/stepDefinitions/mappings/etc. The variables indicate that these are "configs", so I'm wondering if those are the same as "settings".

hao1st replied:
`[
  {
    "name": "CustomerRecords", 
    "config": [
      {
        "name": "Map 1",
        "description": "",
        "sourceFormat": "json",
        "selectedSource":"collection", //(collection  | query)
        "sourceQuery": "",
        "collections": ['Provider'],
        "lastUpdated": "2020-02-15T11:25:35.765299-08:00"
      },
      {
        "name": "Customer Mapping - New",
        "description": "",
        "sourceFormat": "xml",
        "selectedSource":"collection", //(collection  | query)
        "sourceQuery": "cts.uris(null,null,cts.andQuery[cts.CollectionQuery('Customer'),cts.elementAttributeValueQuery()])",
        "collections": ['Customer'],
        "lastUpdated": "2020-02-11T11:25:35.765299-08:00"
      }
    ]
  },
  {
    "name": "StudentCollegeRecordsNew",
    "config": [
      {
        "name": "Map 4",
        "description": "",
        "sourceFormat": "json",
        "selectedSource":"collection", //(collection  | query)
        "sourceQuery": "",
        "collections": ['Provider'],
        "lastUpdated": "2020-02-15T11:25:35.765299-08:00"
      },
      {
        "name": "StudentRec Mapping - New",
        "description": "",
        "sourceFormat": "xml",
        "selectedSource":collection", //(collection  | query),
        "sourceQuery": "",
        "collections": ['Customer'],
        "lastUpdated": "2020-02-11T11:25:35.765299-08:00",
      }
    ]
  }`

hao1st replied:
I guess we had better use "configs" rather than "config" in the json above.

rjrudin replied:
So the use case is "Get me all the artifacts for each known entity type", right? 

If so, shouldn't "configs" be replaced with "artifacts"? You're then returning an array of objects, where each object has an entity type (I recommend replacing "name" with "entityType") and an array of artifacts that reference the entity type.

hao1st replied:
OK, it is fine.

rjrudin commented on marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/artifacts/core.sjs:
```diff
@@ -49,19 +51,31 @@ function getTypesInfo() {
-        for (const result of results) {
-            artifacts.push(result.toObject());
+        if (artifactType == "loadData") {
+            return cts.search(queriesForAnd.length > 1 ? cts.andQuery(queriesForAnd) : queriesForAnd[0]).toArray();
```
One nit - I would just call this "queries" (the plural implying that it's an array). And then it doesn't matter if there's one or more - you can just do:

    cts.search(cts.andQuery(queries)).toArray()

hao1st replied:
Why do you think CollectionsController/Manager should be removed? I noticed that we could not reuse CollectionsService in com.marklogic.hub.web.service package and also depends on HubConfig.

hao1st commented:
@rjrudin make sense. I created a subtask for it.
